src/memory.c

```diff
@@ -411,7 +412,14 @@ static void mmu_add_default_mappings(void)
      * With SPRR enabled, this becomes RW.
      * This range includes all real RAM, including carveouts
      */
-    mmu_add_mapping(ram_base, ram_base, cur_boot_args.mem_size_actual, MAIR_IDX_NORMAL, PERM_RWX);
+
+    /*
+     * Newer machines (m3+) do not always have mem_size_actual filled in.  In this case,
+     * use mem size.  Either way, mask it to a page boundary.
+     */
+    u64 mem_size =
+        cur_boot_args.mem_size_actual ? cur_boot_args.mem_size_actual : cur_boot_args.mem_size;
+    mmu_add_mapping(ram_base, ram_base, mem_size & ~(PAGE_SIZE), MAIR_IDX_NORMAL, PERM_RWX);
```

Comment:
This is not correct. `mem_size` is the usable memory size starting at `phys_base`, not at `ram_base`. `ram_size`, calculated above, is the upper limit of normal RAM starting at RAM base, excluding top carveouts. But even if you used that, you'd still break debugging and scripts which rely on upper RAM being mapped (the reason why we explicitly unmap the TZ carveouts is precisely because mapping *those* is dangerous and causes random faults due to speculation, but mapping everything else in high RAM isn't).

What you want to do is the same thing we do above to cook up `ram_base`: just take the wonky value and round to a multiple of 4GiB. In this case, `mem_size_actual = ALIGN_UP(ram_size, BIT(32))`. This should give real physical RAM size for any reasonable configuration.

I'd go ahead and stick it into `cur_boot_args` too, so we don't have to go around fixing other code that relies on it. If macOS doesn't like that field, I'd rather deal with it in the hypervisor boot sequence for macOS. Just do that at the top of the function if `mem_size_actual` is 0 and leave this line alone :)